velox/py/plan_builder/PyPlanBuilder.h

```diff
@@ -0,0 +1,146 @@
+/// resources and structures needed by Velox.
+void registerAllResources();
+
+class PyVector;
+
+/// Wrapper class to expose Velox plan nodes to Python.
+class PyPlanNode {
+ public:
+  explicit PyPlanNode(core::PlanNodePtr planNode = nullptr)
+      : planNode_(std::move(planNode)) {}
+
+  core::PlanNodePtr getPlanNode() const {
+    return planNode_;
+  }
+
+  std::string id() const {
+    return std::string{planNode_->id()};
+  }
+
+  std::string serialize() const {
+    return folly::toJson(planNode_->serialize());
+  }
+
+  std::string name() const {
+    if (!planNode_) {
+      return "[nullptr]";
+    }
+    return std::string(planNode_->name());
+  }
+
+  std::string toString(bool detailed = false, bool recursive = false) const {
+    if (!planNode_) {
+      return "[nullptr]";
+    }
+    return planNode_->toString(detailed, recursive);
+  }
+
+ private:
+  core::PlanNodePtr planNode_;
+};
+
+/// Wrapper class for PlanBuilder. It allows us to avoid exposing all details of
+/// the class to users making it easier to use.
+class PyPlanBuilder {
+ public:
+  PyPlanBuilder(
+      const std::shared_ptr<core::PlanNodeIdGenerator>& generator = nullptr);
+
+  PyPlanNode getPlanNode() const {
+    return PyPlanNode(planBuilder_.planNode());
+  }
+
+  // Returns a new builder sharing the plan node id generator, such that the new
+  // builder can safely be used to build other parts/pipelines of the same plan.
+  PyPlanBuilder newBuilder() {
+    return PyPlanBuilder{planNodeIdGenerator_};
+  }
+
+  /// Add table scan node with basic functionality.
+  ///
+  /// @param output The schema to be read from the file. Needs to be a RowType.
+  /// @param aliases Aliases to apply, mapping from the desired output name to
+  /// the input name in the file. If there are aliases, OutputSchema should be
+  /// defined based on the aliased names. The Python dict should contain strings
+  /// for keys and values.
+  /// @param subfields Used for feature projection. Dictionary mapping from
+  /// column name (string) to the list of subitems to project from it (list of
+  /// integers - feature ids).
+  /// @param rowIndexColumnName If defined, create an output column with that
+  /// name producing $row_ids. This name needs to be part of `output`.
+  /// @param connectorId ID of the connector to use for this scan.
+  ///
+  /// Example (from Python API):
+  ///
+  /// tableScan(
+  ///   output=ROW(
+  ///     names=["row_number", "$row_group_id", "aliased_name"],
+  ///     types=[BIGINT(), VARCHAR(), MAP(INTEGER(), VARCHAR())],
+  ///   ),
+  ///   subfields={"aliased_name": [1234, 5431, 65223]},
+  ///   row_index="row_number",
+  ///   aliases={
+  ///     "aliased_name": "name_in_file",
+  ///    },
+  ///  )
+  PyPlanBuilder& tableScan(
+      const velox::py::PyType& output,
+      const pybind11::dict& aliases,
+      const pybind11::dict& subfields,
+      const std::string& rowIndexColumnName,
+      const std::string& connectorId);
```

Comment:
Where would users get this ID from?

Reply:
We may need to extend this, but my current thinking is just assume something by default "hive" or "prism", but users can specify an alternative name.

Reply:
(there is a default value in the Python module layer on top of this class). Added a comment to clarify